pkg/ipmi/blob_transfer.go

```diff
@@ -0,0 +1,416 @@
+type CRCOption string
+
+// SessionID is a unique identifier for an open blob.
+type SessionID uint16
+
+// BlobStats contains statistics for a given blob.
+type BlobStats struct {
+	state       uint16
+	size        uint32
+	metadataLen uint8
+	metadata    []uint8
+}
+
+const (
+	IPMI_MAX_PAYLOAD_SIZE = 256
+
+	_IPMI_GGL_NET_FN   = 46
+	_IPMI_GGL_LUN      = 0
+	_IPMI_GGL_BLOB_CMD = 128
+
+	OEN_LEN = 3
+	CRC_LEN = 2
+
+	// Blob transfer command codes
+	BMC_BLOB_CMD_CODE_GET_COUNT    = 0
+	BMC_BLOB_CMD_CODE_ENUMERATE    = 1
+	BMC_BLOB_CMD_CODE_OPEN         = 2
+	BMC_BLOB_CMD_CODE_READ         = 3
+	BMC_BLOB_CMD_CODE_WRITE        = 4
+	BMC_BLOB_CMD_CODE_COMMIT       = 5
+	BMC_BLOB_CMD_CODE_CLOSE        = 6
+	BMC_BLOB_CMD_CODE_DELETE       = 7
+	BMC_BLOB_CMD_CODE_STAT         = 8
+	BMC_BLOB_CMD_CODE_SESSION_STAT = 9
+
+	// Flags
+	BMC_BLOB_OPEN_FLAG_READ  = (1 << 0)
+	BMC_BLOB_OPEN_FLAG_WRITE = (1 << 1)
+	// Blob open: bit positions 2-7 are reserved for future protocol use.
+	// Bit positions 8-15 are available for blob-specific definitions.
+	BMC_BLOB_STATE_OPEN_R       = (1 << 0)
+	BMC_BLOB_STATE_OPEN_W       = (1 << 1)
+	BMC_BLOB_STATE_COMMITTING   = (1 << 2)
+	BMC_BLOB_STATE_COMMITTED    = (1 << 3)
+	BMC_BLOB_STATE_COMMIT_ERROR = (1 << 4)
+	// Blob state: bit positions 5-7 are reserved for future protocol use.
+	// Bit positions 8-16 are available for blob-specific definitions.
+
+	// CRC options
+	REQ_CRC     = "REQ_CRC"
+	RES_CRC     = "RES_CRC"
+	NO_CRC      = "NO_CRC"
+	REQ_RES_CRC = "REQ_RES_CRC"
+
+	// OEM
+	OPENBMC = "OPENBMC"
+)
+
+// Maps OEM names to a 3 byte OEM number.
+// OENs are typically serialized as the first 3 bytes of a request body.
+var OENMap = map[string][3]uint8{
+	OPENBMC: {0xcf, 0xc2, 0x00},
+}
+
+// sendBmcCmd takes a command code, data given in little endian format, and
+// an option for cyclic redundancy checks (CRC). It constructs the request
+// and sends the command over IPMI. It receives the response, validates it,
+// and then returns the response body.
+func (i *IPMI) sendBmcCmd(code uint8, data []uint8, crcOpt CRCOption) ([]byte, error) {
+	// Initialize a buffer with the correct OEN and code.
+	oen, ok := OENMap[OPENBMC]
+	if !ok {
+		return nil, fmt.Errorf("couldn't find OEN for OPENBMC")
+	}
+
+	buf := []uint8{oen[0], oen[1], oen[2], code}
+
+	// If the request should have a CRC, derive a CRC based on the request body.
+	if crcOpt == REQ_CRC || crcOpt == REQ_RES_CRC {
+		crc := new(bytes.Buffer)
+		if err := binary.Write(crc, binary.LittleEndian, genCRC(data)); err != nil {
+			return nil, fmt.Errorf("failed to generate request CRC: ", err)
+		}
+		buf = append(buf, crc.Bytes()...)
+	}
+
+	buf = append(buf, data...)
+
+	// The request buffer should now be as follows:
+	// - 3-byte OEN
+	// - 1-byte subcommand code
+	// - (optionally) 2-byte CRC over request body in little endian format
+	// - request body in little endian format
+
+	req := &req{}
+	addr := systemInterfaceAddr{
```

Comment:
sendrecv already replaces req.addr at all times, right? I don't think this is necessary

Reply:
I changed sendrecv to only replace req.addr if it is nil, so that it was possible to specify the LUN - does that work?

Reply:
yeah